ghcide/src/Development/IDE/GHC/Warnings.hs

```diff
@@ -33,15 +33,12 @@ withWarnings diagSource action = do
   warnings <- newVar []
   let newAction :: DynFlags -> LogActionCompat
       newAction dynFlags logFlags wr _ loc prUnqual msg = do
-        let wr_d = map ((wr,) . third3 (attachReason wr)) $ diagFromErrMsg diagSource dynFlags $ mkWarnMsg dynFlags wr logFlags loc prUnqual msg
+        let wr_d = map ((wr,) . modifyFdLspDiagnostic (attachReason wr)) $ diagFromSDocErrMsg diagSource dynFlags (mkWarnMsg dynFlags wr logFlags loc prUnqual msg)
         modifyVar_ warnings $ return . (wr_d:)
       newLogger env = pushLogHook (const (logActionCompat (newAction (hsc_dflags env)))) (hsc_logger env)
```

Comment:
This whole code is rather suspect. I thought you told me you came up with a way to avoid this entirely?

Reply:
This code is no longer used for anything in the typechecking, but it is still used for bytecode/object code generation, as well as a few other places.

I suspect some of these functions (e.g. codegen) will need deeper changes to be able to get diagnostics as a list, though I don't have great evidence for that atm. I haven't taken a look to see if those functions that are wrapped with this could produce diagnostics another way. I'll take a look another day.

That said, I prefer to defer this to another PR anyways, if at all possible. The current approach gives us structured diagnostics for renaming and typechecking, and doesn't require us to copy too much code from GHC or make any deeper changes, and lets us get started with the bulk of the useful plugin work.

Reply:
Fine with me. If you learned something about it, maybe write it down in a note so we don't forget?

Reply:
I've made a note be37756d